docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -2,25 +2,628 @@
 keywords: [security, concept, authenticity, integrity, certification]
 ---
 
+import TabItem from "@theme/TabItem";
+import { AdornedTabs } from "/src/components/Tabs/AdornedTabs";
+import { AdornedTab } from "/src/components/Tabs/AdornedTab";
+import { BetaChip } from "/src/components/Chip/BetaChip";
 import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
 
 # Security best practices: Data integrity and authenticity
 
-<MarkdownChipRow labels={["Intermediate", "Security", "Concept"]} />
+<MarkdownChipRow labels={["Advanced", "Security", "Concept"]} />
 
-## Certify query responses if they are relevant for security
+## Certified variables
 
 ### Security concern
 
-The responses to [query calls](/docs/current/references/ic-interface-spec#http-interface) are not threshold-signed by the canister and subnet. Thus, a single malicious replica or boundary node may change the data, violating its authenticity. This is especially risky if update calls, which do returned signed responses, depend on the response to query calls.
+ICP offers three modes of operations for canisters, `update`, `query` and `composite_query`. For the sake of simplicity, we will club `composite_query` under queries for the rest of this section. For more information, view the  [detailed overview between `update` and `query` calls](https://internetcomputer.org/docs/current/developer-docs/smart-contracts/call/overview/#introduction). As explained in the overview, update calls are slow and expensive but provide integrity guarantee as they are responded with a threshold signature signed by the subnet. 
 
-### Recommendation
+On the other hand, query calls are fast since they are responded by a single replica but **there is no integrity guarantee, since the response can be manipulated if the responding replica or the boundary node is malicious.** For example, if the NNS dapp fetches proposal information from the governance canister via query calls and the responding node is malicious, it can mask an ill intentioned proposal that causes irrevocable damage as innocuous by modifying the proposal payload in the response and mislead voters into voting yes. Another consequence of query calls is that users can't rely on [canister_inspect_message](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/iam#do-not-rely-on-ingress-message-inspection) as a guard. **This makes query calls, in it's raw form, unfit to serve data for security critical applications.**
```

Comment:
```suggestion
On the other hand, query calls are fast since they are responded by a single replica but **there is no integrity guarantee, since the response can be manipulated by a single replica or the boundary node.** For example, if the NNS dapp fetches proposal information from the governance canister via query calls and the responding node is malicious, it can mask an ill intentioned proposal that causes irrevocable damage as innocuous by modifying the proposal payload in the response and mislead voters into voting yes. Another consequence of query calls is that users can't rely on [canister_inspect_message](https://internetcomputer.org/docs/current/developer-docs/security/security-best-practices/iam#do-not-rely-on-ingress-message-inspection) as a guard. **This makes query calls, in it's raw form, unfit to serve data for security critical applications.**
```